### PR TITLE
add node processor factory

### PR DESCRIPTION
**Pull request by doodzik**
## What are you trying to accomplish?

When I was reading through the code, I got a bit confused about the `RunContext` class's responsibility. The particular issue I saw was that that class is used to build a NodeProcessor and a FileProcessor. This construction is a bit backward because the FileProcessor gets initialized with a RunContext, but is used in the CLI class.

## What approach did you choose, and why?

I did an initial stab at this problem in #44  but gave the CLI class too many responsibilities.
I paired with @exterm on this problem, but we approached it from another side. We decided to call the FillProcessor#call method from the RunContext class instead of exposing the FileProcessor class to the CLI class. 

After that, I changed the code to pass a NodeProcessorFactory to the
FileProcessor instead of a RunContext instance, which reduces how complected these classes are. 

There are some other minor cleanups in the PR, but those are described in the commit messages.

## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

### Additional Release Notes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Include any notes here to include in the release description. For example, if you selected "breaking change" above, leave notes on how users can transition to this version.

If no additional notes are necessary, delete this section or leave it unchanged.

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] It is safe to rollback this change.

### CODE REVIEW

**Review by rochlefebvre — APPROVED** [review]
The changes make sense, according to my limited experience with Packwerk.  The new methods in RunContext clean up the initializer quite nicely. I suggest you wait for a second review from someone with more Packwerk domain knowledge.

**Review by tomstuart — COMMENTED** [review]
I’ve run the `node_processor_factory` branch against the main Shopify monolith and confirmed that its `packwerk check` and `packwerk update` behaviour is the same as on its current base commit (dac7b105). There are merge conflicts against `main` so I haven’t yet tried this with a merged/rebased version.

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb:16]: I assume this `require` belongs in a later commit, since this file doesn’t exist (yet).

File: lib/packwerk/file_processor.rb
Comment (tomstuart) [inline lib/packwerk/file_processor.rb:35]: It’s interesting that we had an unnecessary instance variable here. Looking at the history, it was originally used by a `#visit` method; that method got removed when we extracted `Packwerk::NodeVisitor` but the instance variable was left behind.

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb:38]: Good catch!

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb:59]: It took me a little while to understand why this is okay, but I think it’s because a) the only place where `RunContext` gets constructed is in `RunContext.from_configuration`, which uses `Configuration#root_path`, and b) `Configuration` already calls `File.expand_path` in its constructor so `Configuration#root_path` has always been expanded.

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb:29]: 👀

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb:122]: I agree we should do this, but there are still references to `@load_paths`, `@package_paths`, `@inflector` and `@checker_classes` throughout the class — should we clean those up too?

File: lib/packwerk/node_processor_factory.rb
Comment (tomstuart) [inline lib/packwerk/node_processor_factory.rb:19]: This is good I think. I find it easier to understand this way, and if we’re not losing any capability then that seems like a win. As always, we can easily switch back to injecting the class if/when we ever need it.

File: lib/packwerk/cli.rb
Comment (tomstuart) [inline lib/packwerk/cli.rb:26]: Nice! I think we should add a `require "packwerk/checking_deprecated_references"` in this file now.

File: lib/packwerk/run_context.rb
Comment (tomstuart) [inline lib/packwerk/run_context.rb]: I think you can remove `require "packwerk/checking_deprecated_references"` now.

**Review by tomstuart — APPROVED** [review]
This looks good, thank you @doodzik! I left a few suggestions but didn’t see any blockers.

File: lib/packwerk/run_context.rb
Comment (doodzik) [reply lib/packwerk/run_context.rb:16]: Nice catch, I probably messed up the `git commit -p`.

File: test/unit/cli_test.rb
Comment (doodzik) [reply test/unit/cli_test.rb:49]: I changed it to `at_least(2)` since that is more accurate. Good catch 👍

**Comment by doodzik** [comment]
I fixed the merge conflict and ran the new version of packwerk on core.

**Pull request merged**
